vortex-array/src/arrays/decimal/mod.rs

```diff
@@ -72,30 +78,47 @@ pub struct DecimalArray {
 }
 
 impl DecimalArray {
-    /// Creates a new [`DecimalArray`] from a [`Buffer`] and [`Validity`], without checking
-    /// any invariants.
-    pub fn new<T: NativeDecimalType>(
+    /// Creates a new [`DecimalArray`] from a [`Buffer`] and [`Validity`].
+    pub fn try_new<T: NativeDecimalType>(
         buffer: Buffer<T>,
         decimal_dtype: DecimalDType,
         validity: Validity,
-    ) -> Self {
+    ) -> VortexResult<Self> {
         if let Some(len) = validity.maybe_len() {
             if buffer.len() != len {
-                vortex_panic!(
+                vortex_bail!(
                     "Buffer and validity length mismatch: buffer={}, validity={}",
                     buffer.len(),
                     len,
                 );
             }
         }
 
-        Self {
+        if !compatible_storage_type(T::VALUES_TYPE, decimal_dtype) {
+            vortex_bail!(
+                "{:?} cannot represent every value in {}",
+                T::VALUES_TYPE,
+                decimal_dtype
+            )
+        }
```

Comment:
I guess we can assert that in the patch. T could be smaller than the smallests possible width for dtype